src/validation/recursive_validator.rs

```diff
@@ -113,11 +104,31 @@ impl RecursiveValidator {
                     .collect();
 
                 slice.push(node); // Append to get `B -> C -> B` instead of `B -> C` in the report
-
                 self.diagnostics.push(Diagnostic::recursive_datastructure(&slice.join(" -> "), ranges));
             }
 
             None => unreachable!("Node has to be in the IndexSet"),
         }
     }
+
+    /// Returns the type name of `entry` distinguishing between two cases:
+    /// 1. If the entry is any type but an array its datatype is returned (as usual)
+    /// 2. If the entry is an array their inner type name is returned. For example calling the
+    /// [`index::VariableIndexEntry::get_type_name`] method on the following code snippet
+    /// ```ST
+    /// TYPE A : STRUCT
+    ///     b : ARRAY[0..1] OF B;
+    /// END_STRUCT END_TYPE
+    /// ```
+    /// would return `__A_b` as the datatype for `b` whereas it should have been `B`, i.e. their inner type
+    /// name. This is important for the `dfs` method as it otherwise wouldn't correctly recognize cycles since
+    /// it operate on these "normalized" type names.
+    #[inline(always)]
+    fn get_type_name<'idx>(&self, index: &'idx Index, entry: &'idx VariableIndexEntry) -> &'idx str {
+        let type_name = entry.get_type_name();
+        match index.get_type_information_or_void(type_name).get_inner_array_type_name() {
```

Comment:
This feels sneaky to me, is it better to use the is_array method you added here instead of get_inner_array_type_name returning none?
From the readability point of view, this reads as "I have an array but it has no type name", not "I have no array so I use the normal type name"

Reply:
Done. I agree with you in that using `is_array` makes the function more readable.